url24d793285d_auzll_gengweibo_tgz-pJ8-org_json_TestJ8/src/net/oauth/OAuth.java

```diff
@@ -0,0 +1,381 @@
+            map.put(Problems.TOKEN_REVOKED, unauthorized);
+            map.put(Problems.TOKEN_REJECTED, unauthorized);
+            map.put("token_not_authorized", unauthorized);
+            map.put(Problems.SIGNATURE_INVALID, unauthorized);
+            map.put(Problems.CONSUMER_KEY_UNKNOWN, unauthorized);
+            map.put(Problems.CONSUMER_KEY_REJECTED, unauthorized);
+            map.put(Problems.ADDITIONAL_AUTHORIZATION_REQUIRED, unauthorized);
+            map.put(Problems.PERMISSION_UNKNOWN, unauthorized);
+            map.put(Problems.PERMISSION_DENIED, unauthorized);
+
+            map.put(Problems.USER_REFUSED, serviceUnavailable);
+            map.put(Problems.CONSUMER_KEY_REFUSED, serviceUnavailable);
+            return Collections.unmodifiableMap(map);
+        }
+
+    }
+
+    private static String characterEncoding = ENCODING;
+
+    public static void setCharacterEncoding(String encoding) {
+	OAuth.characterEncoding = encoding;
+    }
+
+    public static String decodeCharacters(byte[] from) {
+	if (characterEncoding != null) {
+	    try {
+		return new String(from, characterEncoding);
+	    } catch (UnsupportedEncodingException e) {
+		System.err.println(e + "");
+	    }
+	}
+	return new String(from);
+    }
+
+    public static byte[] encodeCharacters(String from) {
+	if (characterEncoding != null) {
+	    try {
+		return from.getBytes(characterEncoding);
+	    } catch (UnsupportedEncodingException e) {
+		System.err.println(e + "");
+	    }
+	}
+	return from.getBytes();
+    }
+
+    /** Return true if the given Content-Type header means FORM_ENCODED. */
+    public static boolean isFormEncoded(String contentType) {
+        if (contentType == null) {
+            return false;
+        }
+        int semi = contentType.indexOf(";");
+        if (semi >= 0) {
+            contentType = contentType.substring(0, semi);
+        }
+        return FORM_ENCODED.equalsIgnoreCase(contentType.trim());
+    }
+
+    /**
+     * Construct a form-urlencoded document containing the given sequence of
+     * name/value pairs. Use OAuth percent encoding (not exactly the encoding
+     * mandated by HTTP).
+     */
+    public static String formEncode(Iterable<? extends Map.Entry> parameters)
+            throws IOException {
+        ByteArrayOutputStream b = new ByteArrayOutputStream();
+        formEncode(parameters, b);
+        return decodeCharacters(b.toByteArray());
+    }
+
+    /**
+     * Write a form-urlencoded document into the given stream, containing the
+     * given sequence of name/value pairs.
+     */
+    public static void formEncode(Iterable<? extends Map.Entry> parameters,
+            OutputStream into) throws IOException {
+        if (parameters != null) {
+            boolean first = true;
+            for (Map.Entry parameter : parameters) {
+                if (first) {
+                    first = false;
+                } else {
+                    into.write('&');
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;into&#39; (index: 1 | type: OutputStream). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;parameters&#39;. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).